ondevice_learning.qmd

```diff
@@ -1,61 +1,630 @@
+
+For instance, a cloud model can accurately detect pneumonia in chest X-rays from thousands of hospitals. However, an on-device model trained only on a small local patient population may fail to generalize.
+
+Unreliable accuracy limits the real-world applicability of on-device learning for mission-critical uses like disease diagnosis or self-driving vehicles.
+
+On-device training is also slower than the cloud due to limited resources. Even if each iteration is faster, the overall training process takes longer. 
+
+For example, a real-time robotics application may require model updates within milliseconds. On-device training on small embedded hardware may take seconds or minutes per update - too slow for real-time use.
+
+Accuracy, generalization, and speed challenges pose hurdles to adopting on-device learning for real-world production systems, especially when reliability and low latency are critical.
+
+
+## Ondevice Adaptation
+
+In a machine learning task, resource consumption mainly comes from three sources: 
+
+* The ML model itself; 
+* The optimization process during model learning
+* Storing and processing the dataset used for learning. 
+
+Correspondingly, there are three approaches to adapting existing ML algorithms onto resource-constrained devices: 
+
+* Reducing the complexity of the ML model
+* Modifying optimizations to reduce training resource requirements
+* Creating new storage-efficient data representations
+
+In the following section, we will review these on-device learning adaptation methods. More details on model optimizations can be found in the [Model Optimizations](./optimizations.qmd) chapter. 
+
+### Reducing Model Complexity
+
+In this section, we will briefly discuss ways to reduce model complexity to adapt ML models on-device. For details of reducing model complexity, please refer to the Model Optimization Chapter. 
+
+#### Traditional ML Algorithms
+
+Due to the compute and memory limitations of edge devices, select traditional ML algorithms are great candidates for on-device learning applications due to their lightweight nature. Some example algorithms with low resource footprints include Naive Bayes Classifier, Support Vector Machines (SVMs), Linear Regression, Logistic Regression, and select Decision Tree algorithms. 
+
+With some refinements, these classical ML algorithms can be adapted to specific hardware architectures and perform simple tasks, and their low performance requirements make it easy to integrate continuous learning even on edge devices. 
+
+#### Pruning
+
+Pruning is a technique used to reduce the size and complexity of a decision tree or other types of machine learning models, like neural networks, to improve their efficiency and generalization performance. This is beneficial for training models on edge devices, where we want to minimize the resource usage while maintaining competitive accuracy. 
```

Comment:
I would remove the mention of decision trees and just mention pruning as a general technique for ML models.

Reply:
Addressed

Reply:
done